packages/shared/components/FileTransfer/FileTransfer.tsx

```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright 2022 Gravitational, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+
+import { useFileTransferContext } from './FileTransferContextProvider';
+import { useFilesStore } from './useFilesStore';
+import {
+  FileTransferDialogDirection,
+  FileTransferStateless,
+  RunFileTransfer,
+} from './FileTransferStateless';
+
+interface FileTransferProps {
+  backgroundColor?: string;
+  transferHandlers: TransferHandlers;
+
+  /**
+   * `beforeClose` is called when an attempt to close the dialog was made
+   * and there is a file transfer in progress.
+   * Returning `true` will close the dialog, returning `false` will not.
+   */
+  beforeClose?(): Promise<boolean> | boolean;
+
+  afterClose?(): void;
+}
+
+export interface TransferHandlers {
+  getDownloader: (sourcePath: string) => Promise<RunFileTransfer | undefined>;
+  getUploader: (
+    destinationPath: string,
+    file: File
+  ) => Promise<RunFileTransfer | undefined>;
+}
+
+export function FileTransfer(props: FileTransferProps) {
+  const { openedDialog, closeDialog } = useFileTransferContext();
+
+  async function handleCloseDialog(
+    isAnyTransferInProgress: boolean
+  ): Promise<void> {
+    const runCloseCallbacks = () => {
+      closeDialog();
+      props.afterClose?.();
+    };
+
+    if (!isAnyTransferInProgress) {
+      runCloseCallbacks();
+      return;
+    }
+
+    if (props.beforeClose ? await props.beforeClose() : true) {
+      runCloseCallbacks();
+    }
```

Comment:
If there's no `beforeClose`, we want to always just run the close callbacks, right?

How about this:

```suggestion
    if (!isAnyTransferInProgress || !props.beforeClose) {
      runCloseCallbacks();
      return;
    }

    if (await props.beforeClose()) {
      runCloseCallbacks();
    }
```

This way we don't have to deal with that ternary branch that always returns true.

Reply:
Thx, I forgot to simplify this logic :) Done